test/common/memory/memory_release_test.cc

```diff
@@ -0,0 +1,143 @@
+#include "source/common/event/dispatcher_impl.h"
+#include "source/common/memory/stats.h"
+
+#include "test/common/stats/stat_test_utility.h"
+#include "test/test_common/simulated_time_system.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+
+namespace Envoy {
+namespace Memory {
+
+class AllocatorManagerPeer {
+public:
+  static std::chrono::milliseconds
+  memoryReleaseInterval(const AllocatorManager& allocator_manager) {
+    return allocator_manager.memory_release_interval_msec_;
+  }
+  static uint64_t bytesToRelease(const AllocatorManager& allocator_manager) {
+    return allocator_manager.bytes_to_release_;
+  }
+};
+
+namespace {
+
+static const int MB = 1048576;
+
+class MemoryReleaseTest : public testing::Test {
+protected:
+  MemoryReleaseTest()
+      : api_(Api::createApiForTest(stats_, time_system_)),
+        dispatcher_("test_thread", *api_, time_system_), scope_("memory_release_test.", stats_) {}
+
+  void initialiseAllocatorManager(uint64_t bytes_to_release, float release_interval_s) {
+    const std::string yaml_config = (release_interval_s > 0)
+                                        ? fmt::format(R"EOF(
+  bytes_to_release: {}
+  memory_release_interval: {}s
+)EOF",
+                                                      bytes_to_release, release_interval_s)
+                                        : fmt::format(R"EOF(
+  bytes_to_release: {}
+)EOF",
+                                                      bytes_to_release);
+    const auto proto_config =
+        TestUtility::parseYaml<envoy::config::bootstrap::v3::MemoryAllocatorManager>(yaml_config);
+    allocator_manager_ = std::make_unique<Memory::AllocatorManager>(*api_, scope_, proto_config);
+  }
+
+  void step(const std::chrono::milliseconds& step) { time_system_.advanceTimeWait(step); }
+
+  Envoy::Stats::TestUtil::TestStore stats_;
+  Event::SimulatedTimeSystem time_system_;
+  Api::ApiPtr api_;
+  Event::DispatcherImpl dispatcher_;
+  Envoy::Stats::TestUtil::TestScope scope_;
+  std::unique_ptr<Memory::AllocatorManager> allocator_manager_;
+};
+
+TEST_F(MemoryReleaseTest, ReleaseRateAboveZeroDefaultIntervalMemoryReleased) {
+  size_t initial_allocated_bytes = Stats::totalCurrentlyAllocated();
+  auto a = std::make_unique<unsigned char[]>(MB);
+  auto b = std::make_unique<unsigned char[]>(MB);
+  if (Stats::totalCurrentlyAllocated() - initial_allocated_bytes <= 0) {
+    // Cannot measure memory usage precisely on this platform.
```

Comment:
Might be useful to gtest skip e.g. so we can see when the test cannot run on a platform vs the test fully ran and passed: https://google.github.io/googletest/advanced.html#skipping-test-execution